src/Channels/Memory.cs

```diff
@@ -0,0 +1,90 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+
+namespace Channels
+{
+    public struct Memory<T>
+    {
+        private readonly T[] _array;
+        private readonly int _offset;
+        private readonly unsafe void* _memory;
+        private readonly int _memoryLength;
+
+        public unsafe Memory(void* pointer, int offset, int length)
+        {
+            unsafe
+            {
+                _memory = pointer;
+            }
+
+            _array = null;
+            _offset = offset;
+            _memoryLength = length;
+        }
+
+        public unsafe Memory(T[] array, int offset, int length, void* pointer = null)
+        {
+            unsafe
+            {
+                _memory = pointer;
+            }
+
+            _array = array;
+            _offset = offset;
+            _memoryLength = length;
+        }
+
+        public Span<T> Span => this;
+
+        public bool IsEmpty => Length == 0;
+
+        public static implicit operator Span<T>(Memory<T> memory)
+        {
+            if (memory.Length == 0)
+            {
+                return Span<T>.Empty;
+            }
+
+            if (memory._array != null)
+            {
+                return memory._array.Slice(memory._offset, memory.Length);
+            }
+            else
+            {
+                unsafe
+                {
+                    return new Span<T>(memory._memory, memory._memoryLength);
+                }
+            }
+        }
+
+        public unsafe void* UnsafePointer => (byte*)_memory + (Unsafe.SizeOf<T>() * _offset);
```

Comment:
As a corollary of that, I wonder if the constructor parameters should actually be `bool isPinned` rather than passing in the pointer. That way you can't get wildly inconsistent values from the array and the pointer. If they tell us it is pinned, the code could store the address via `fixed`. If they lied and it isn't pinned: that's their fault.